mypy/typeops.py

```diff
@@ -130,6 +145,18 @@ def map_type_from_supertype(typ: Type,
     return expand_type_by_instance(typ, inst_type)
 
 
+def supported_self_type(typ: ProperType) -> bool:
+    """Is this a supported kind of explicit self-types?
+
+    Currently, this means a X or Type[X], where X is an instance or
+    a type variable with an instance upper bound.
+    """
+    if isinstance(typ, TypeType):
+        return supported_self_type(typ.item)
+    return (isinstance(typ, TypeVarType) or
```

Comment:
This doesn't seem to check that there is an instance upper bound? Should there be?

Also my inclination would be to throw parens around the and clause.

Reply:
> This doesn't seem to check that there is an instance upper bound? Should there be?

I would say it is safe to keep it as is. We can restrict this later if we will find problems.
> Also my inclination would be to throw parens around the and clause.

OK.